### PR TITLE
Be more precise about what max/minlength measures

### LINKED ISSUE #23783: Inconsistencies in character length definition.
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/maxlength

### What specific section or headline is this issue about?

HTML attribute: maxlength

### What information was incorrect, unhelpful, or incomplete?

The HTML maxLength documentation says that characters are code units.

> The maxlength attribute defines the maximum number of characters (as UTF-16 code units) the user can enter into an [<input>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) or [<textarea>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea).

### What did you expect to see?

The HTML doc should follow the javascript length doc and define "code units" separately from "characters"

### Do you have any supporting links, references, or citations?

The `String length` [javascript documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length) differentiates between "code units" and "characters"

> Since length counts code units instead of characters, if you want to get the number of characters, you can first split the string with its [iterator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/@@iterator), which iterates by characters:

### Do you have anything more you want to share?

_No response_

### PR DESCRIPTION
**Pull request by Josh-Cena**
### Description

"Character" usually refers to Unicode characters, so saying "character as UTF-16 code units" is a bit of an oxymoron. This PR makes the wording technically more correct.

Fixes https://github.com/mdn/content/issues/23783

### CODE REVIEW

File: files/en-us/web/html/attributes/maxlength/index.md [hunk 1]
```diff
@@ -9,7 +9,7 @@ browser-compat:
 
 {{HTMLSidebar}}
 
-The **`maxlength`** attribute defines the maximum number of characters (as UTF-16 code units) the user can enter into an {{htmlelement('input')}} or {{htmlelement('textarea')}}. This must be an integer value 0 or higher.
+The **`maxlength`** attribute defines the maximum string length the user can enter into an {{htmlelement('input')}} or {{htmlelement('textarea')}}. This must be an integer value 0 or higher. The string [length](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length) is measured in UTF-16 code units, which is usually equivalent to the number of characters for most scripts, although not always — [Strings with length not equal to the number of characters](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length#strings_with_length_not_equal_to_the_number_of_characters) has some examples.
 
 If no maxlength is specified, or an invalid value is specified, the input or textarea has no maximum length.
```
Comment (hamishwillee) [inline files/en-us/web/html/attributes/maxlength/index.md:14]: I love the precision. I have preference for early forward links and having a very short introductory sentence. I also don't feel the need to fully explain where I'm linking to if the likely content of the link can be inferred. This is just to my taste though:

```suggestion
The **`maxlength`** attribute defines the maximum [string length](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length) that the user can enter into an {{htmlelement('input')}} or {{htmlelement('textarea')}}.

The attribute must be an integer value of 0 or higher. The length is measured in UTF-16 code units, which ([for most scripts](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length#strings_with_length_not_equal_to_the_number_of_characters)) are equivalent to the number of characters.
If no maxlength is specified, or an invalid value is specified, the `<input>` or `<textarea>` has no maximum length.
```

Comment (hamishwillee) [reply files/en-us/web/html/attributes/maxlength/index.md:14]: If you don't agree, just fix up the highlighting

```suggestion
The **`maxlength`** attribute defines the maximum string length the user can enter into an {{htmlelement('input')}} or {{htmlelement('textarea')}}. This must be an integer value of 0 or higher. The string [length](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length) is measured in UTF-16 code units, which is usually equivalent to the number of characters for most scripts, although not always — [Strings with length not equal to the number of characters](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length#strings_with_length_not_equal_to_the_number_of_characters) has some examples.

If no `maxlength` is specified, or an invalid value is specified, the `<input>` or `<textarea>` has no maximum length.
```

File: files/en-us/web/html/element/input/email/index.md [hunk 2]
```diff
@@ -33,13 +33,13 @@ The values of the list attribute is the {{domxref("Element.id", "id")}} of a {{H
 
 ### maxlength
 
-The maximum number of characters (as UTF-16 code units) the user can enter into the `email` input. This must be an integer value 0 or higher. If no `maxlength` is specified, or an invalid value is specified, the `email` input has no maximum length. This value must also be greater than or equal to the value of `minlength`.
+The maximum string length (measured in UTF-16 code units) the user can enter into the `email` input. This must be an integer value 0 or higher. If no `maxlength` is specified, or an invalid value is specified, the `email` input has no maximum length. This value must also be greater than or equal to the value of `minlength`.
```
Comment (hamishwillee) [inline files/en-us/web/html/element/input/email/index.md:36]:
```suggestion
The maximum string length (measured in UTF-16 code units) that the user can enter into the `email` input. This must be an integer value of 0 or higher. If no `maxlength` is specified, or an invalid value is specified, the `email` input has no maximum length. This value must also be greater than or equal to the value of `minlength`.
```

File: files/en-us/web/html/element/input/email/index.md [hunk 3]
```diff
@@ -33,13 +33,13 @@ The values of the list attribute is the {{domxref("Element.id", "id")}} of a {{H
...
 
 The input will fail [constraint validation](/en-US/docs/Web/HTML/Constraint_validation) if the length of the text value of the field is greater than `maxlength` UTF-16 code units long. Constraint validation is only applied when the value is changed by the user.
 
 ### minlength
 
-The minimum number of characters (as UTF-16 code units) the user can enter into the `email` input. This must be a non-negative integer value smaller than or equal to the value specified by `maxlength`. If no `minlength` is specified, or an invalid value is specified, the `email` input has no minimum length.
+The minimum string length (measured in UTF-16 code units) the user can enter into the `email` input. This must be a non-negative integer value smaller than or equal to the value specified by `maxlength`. If no `minlength` is specified, or an invalid value is specified, the `email` input has no minimum length.
```
Comment (hamishwillee) [inline files/en-us/web/html/element/input/email/index.md:42]:
```suggestion
The minimum string length (measured in UTF-16 code units) that the user can enter into the `email` input. This must be a non-negative integer value smaller than or equal to the value specified by `maxlength`. If no `minlength` is specified, or an invalid value is specified, the `email` input has no minimum length.
```

File: files/en-us/web/html/element/input/index.md [hunk 4]
```diff
@@ -480,7 +480,7 @@ A few additional non-standard attributes are listed following the descriptions o
 
 - `minlength`
 
-  - : Valid for `text`, `search`, `url`, `tel`, `email`, and `password`, it defines the minimum number of characters (as UTF-16 code units) the user can enter into the entry field. This must be a non-negative integer value smaller than or equal to the value specified by `maxlength`. If no `minlength` is specified, or an invalid value is specified, the input has no minimum length.
+  - : Valid for `text`, `search`, `url`, `tel`, `email`, and `password`, it defines the minimum string length (measured in UTF-16 code units) the user can enter into the entry field. This must be a non-negative integer value smaller than or equal to the value specified by `maxlength`. If no `minlength` is specified, or an invalid value is specified, the input has no minimum length.
```
Comment (hamishwillee) [inline files/en-us/web/html/element/input/index.md:483]:
```suggestion
  - : Valid for `text`, `search`, `url`, `tel`, `email`, and `password`, it defines the minimum string length (measured in UTF-16 code units) that the user can enter into the entry field. This must be a non-negative integer value smaller than or equal to the value specified by `maxlength`. If no `minlength` is specified, or an invalid value is specified, the input has no minimum length.
```

File: files/en-us/web/html/element/input/password/index.md [hunk 5]
```diff
@@ -35,13 +35,13 @@ In addition to the attributes that operate on all {{HTMLElement("input")}} eleme
 
 ### maxlength
 
-The maximum number of characters (as UTF-16 code units) the user can enter into the password field. This must be an integer value 0 or higher. If no `maxlength` is specified, or an invalid value is specified, the password field has no maximum length. This value must also be greater than or equal to the value of `minlength`.
+The maximum string length (measured in UTF-16 code units) the user can enter into the password field. This must be an integer value 0 or higher. If no `maxlength` is specified, or an invalid value is specified, the password field has no maximum length. This value must also be greater than or equal to the value of `minlength`.
```
Comment (hamishwillee) [inline files/en-us/web/html/element/input/password/index.md:38]:
```suggestion
The maximum string length (measured in UTF-16 code units) that the user can enter into the password field. This must be an integer value of 0 or higher. If no `maxlength` is specified, or an invalid value is specified, the password field has no maximum length. This value must also be greater than or equal to the value of `minlength`.
```

File: files/en-us/web/html/element/input/index.md [hunk 6]
```diff
@@ -466,7 +466,7 @@ A few additional non-standard attributes are listed following the descriptions o
 
 - `maxlength`
 
-  - : Valid for `text`, `search`, `url`, `tel`, `email`, and `password`, it defines the maximum number of characters (as UTF-16 code units) the user can enter into the field. This must be an integer value `0` or higher. If no `maxlength` is specified, or an invalid value is specified, the field has no maximum length. This value must also be greater than or equal to the value of `minlength`.
+  - : Valid for `text`, `search`, `url`, `tel`, `email`, and `password`, it defines the maximum string length (measured in UTF-16 code units) the user can enter into the field. This must be an integer value `0` or higher. If no `maxlength` is specified, or an invalid value is specified, the field has no maximum length. This value must also be greater than or equal to the value of `minlength`.
```
Comment (hamishwillee) [inline files/en-us/web/html/element/input/index.md:469]: Note, I prefer the `0` but removed for consistency.

```suggestion
  - : Valid for `text`, `search`, `url`, `tel`, `email`, and `password`, it defines the maximum string length (measured in UTF-16 code units) that the user can enter into the field. This must be an integer value of 0 or higher. If no `maxlength` is specified, or an invalid value is specified, the field has no maximum length. This value must also be greater than or equal to the value of `minlength`.
```

File: files/en-us/web/html/element/input/email/index.md (hunk 2 above)
Comment (hamishwillee) [reply files/en-us/web/html/element/input/email/index.md:36]: PS I added that ^^^ in a lot of places, but not all. 

We mention UTF code units lots, but you defined once. You might want to cross link or glossary. Up to you.

**Comment by Josh-Cena** [comment]
@hamishwillee Thanks for the suggestions—great catches. I think linking to UTF-16 from the attribute pages is good enough.

**Review by hamishwillee — APPROVED** [review]
>  I think linking to UTF-16 from the attribute pages is good enough.

I think that if you're wrong, with the work you have done we are in a very good place to fix this in future :-). Either way, this is a heap better than it was.

Thanks very much.

**Pull request merged**